controller/system_backup_controller.go

```diff
@@ -0,0 +1,1068 @@
+
+	case longhorn.SystemBackupStateUpload:
+		go c.UploadSystemBackup(systemBackup, archievePath, tempDir, backupTargetClient)
+
+	case longhorn.SystemBackupStateReady:
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+	case longhorn.SystemBackupStateError:
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+	case longhorn.SystemBackupStateDelete:
+		err = c.cleanupRemoteSystemBackupFiles(systemBackup, backupTargetClient)
+		if err != nil {
+			c.cacheStateUpdate(
+				systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+				longhorn.SystemBackupConditionReasonDelete, SystemBackupErrDelete,
+			)
+			return err
+		}
+
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+		err = c.ds.RemoveFinalizerForSystemBackup(systemBackup)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *SystemBackupController) InitSystemBackup(systemBackup *longhorn.SystemBackup, log logrus.FieldLogger) error {
+	systemBackup.Status.Version = meta.Version
+	systemBackup.Status.GitCommit = meta.GitCommit
+	return nil
+}
+
+func (c *SystemBackupController) UploadSystemBackup(systemBackup *longhorn.SystemBackup, archievePath, tempDir string, backupTargetClient engineapi.BackupTargetClientSystemBackupInterface) {
+	log := getLoggerForSystemBackup(c.logger, systemBackup.Name, backupTargetClient.GetURL())
+
+	var recordErr error
+	existingSystemBackup := systemBackup.DeepCopy()
+	defer func() {
+		if recordErr != nil {
+			c.cacheStateUpdate(
+				systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+				longhorn.SystemBackupConditionReasonUpload, recordErr.Error(),
+			)
+		} else {
+			systemBackup.Status.ManagerImage = c.managerImage
+			systemBackup.Status.CreatedAt = metav1.Time{Time: time.Now().UTC()}
+
+			c.cacheStateUpdate(
+				systemBackupRecordTypeNormal, longhorn.SystemBackupStateReady,
+				longhorn.SystemBackupConditionReasonUpload, SystemBackupMsgUploadBackupTarget,
+			)
+		}
+
+		c.handleState(systemBackup, existingSystemBackup, recordErr, log)
+	}()
+
+	defaultEngineImage, err := c.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
+	if err != nil {
+		recordErr = errors.Wrapf(err, SystemBackupErrGetFmt, "default engine image")
+		return
+	}
+
+	timeout := time.Duration(datastore.SystemBackupTimeout) * time.Second
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	var systemBackupCfg *systembackupstore.Config
+	for {
+		select {
+		case <-timer.C:
+			recordErr = errors.Wrap(err, SystemBackupErrTimeoutUpload)
+			return
+		case <-ticker.C:
+			systemBackupCfg, err = backupTargetClient.SystemBackupConfigGet(systemBackup.Name, systemBackup.Status.Version)
+			if err != nil && !types.ErrorIsNotFound(err) {
+				log.WithError(err).Debugf("Getting system backup config")
+				err = errors.Wrap(err, SystemBackupErrGetConfig)
+				continue
+			}
+
+			if systemBackupCfg != nil {
+				return
+			}
```

Comment:
Is the cfg file generation & uploading the last step?

Reply:
The intention of this check is to have the check wrapped around the upload:
1. Do not upload if somehow the file already exists in the backup target.
2. Checks the existence of the file in the backup target after the upload then break.